presto-hive/src/main/java/com/facebook/presto/hive/s3/PrestoS3FileSystem.java

```diff
@@ -1288,25 +1349,97 @@ public void close()
 
             try {
                 super.close();
-                uploadObject();
+                // if the upload wasn't initiated already, use the AWS transfer manager
+                // otherwise, flush the final part and upload complete the multipart request
+                String etag = upload.map(upload -> {
+                    flushPartUpload(true);
+                    return client.completeMultipartUpload(new CompleteMultipartUploadRequest()
+                            .withBucketName(host)
+                            .withKey(key)
+                            .withPartETags(partList)
+                            .withUploadId(upload.getUploadId())).getETag();
+                }).orElseGet(this::uploadSingleFile);
+                log.debug("Uploaded file to %s with ETag: %s", key, etag);
             }
             finally {
-                if (!tempFile.delete()) {
-                    log.warn("Could not delete temporary file: %s", tempFile);
+                try (Stream<java.nio.file.Path> dirStream = Files.walk(tempDir.toPath())) {
+                    if (!dirStream.map(java.nio.file.Path::toFile)
+                            // deletes all files first, then the directory
+                            .sorted(Comparator.reverseOrder())
+                            .map(File::delete)
+                            .reduce(true, (a, b) -> a && b)) {
+                        log.warn("Failed to delete all files in " + tempDir);
+                    }
                 }
-                // close transfer manager but keep underlying S3 client open
-                transferManager.shutdownNow(false);
             }
         }
 
-        private void uploadObject()
-                throws IOException
+        private void flushPartUpload(boolean last)
```

Comment:
With this implementation, we are not utilizing the disk as a buffer adequately IMO. The uploads follow a create-part->upload-part-> delete-part flow, which means that our overall write speed is bounded by the network speed/bw (or the disk speed).
We could address this by creating a pipelined writer where the upload and delete of a part happens parallelly in the background and while the part-writer produces parts in a different thread. We can cap the total number of on-disk parts so that we don't fill up the disk.

Reply:
I had considered something like this as an option. However, I felt the pipelined version was too complex to pursue. The main goal for this PR was just to lift the disk-size limitation. Additional performance gain was a non-goal, but we happened to get some by properly overriding the `FilterOutputStream`

Reply:
With the results [here](https://github.com/prestodb/presto/pull/22424#issuecomment-2052150813), I am of the opinion that we should let this PR merge since the performance is always better than master in the benchmark configurations. If we want to improve perf further, we can iterate in another PR on a proper streaming implementation. Trino also has a proper streaming implementation.

When implementing the pipelining I think there are more design decisions that we may wish to make (or further configuration) which is not really in scope for the goal of the PR, which was lifting the disk space limitation.

Reply:
Sounds good. Can you make an issue and link as a TODO for the streaming implementation ?

Reply:
Created https://github.com/prestodb/presto/issues/22536